File: VizOs2/VizOs/frontend/js/priority.js
// Priority CPU Scheduling Algorithm
class Priority {
    static simulate(processes) {
        // Sort processes by arrival time first
        const sortedProcesses = [...processes].sort((a, b) => a.arrival - b.arrival);
        
        let currentTime = 0;
        const ganttChart = [];
        const processResults = [];
        const readyQueue = [];
        let processIndex = 0;
        
        while (processIndex < sortedProcesses.length || readyQueue.length > 0) {
            // Add processes that have arrived to ready queue
            while (processIndex < sortedProcesses.length && 
                   sortedProcesses[processIndex].arrival <= currentTime) {
                readyQueue.push(sortedProcesses[processIndex]);
                processIndex++;
            }
            
            if (readyQueue.length === 0) {
                // No processes ready, advance time to next arrival
                if (processIndex < sortedProcesses.length) {
                    currentTime = sortedProcesses[processIndex].arrival;
                    continue;
                } else {
                    break;
                }
            }
            
            // Sort ready queue by priority (lower number = higher priority)
            readyQueue.sort((a, b) => a.priority - b.priority);
            
            // Execute the highest priority job
            const process = readyQueue.shift();
            const startTime = Math.max(currentTime, process.arrival);
            const completionTime = startTime + process.burst;
            const turnaroundTime = completionTime - process.arrival;
            const waitingTime = turnaroundTime - process.burst;
            const responseTime = startTime - process.arrival;
            
            // Add to Gantt chart
            ganttChart.push({
                name: process.id,
                startTime: startTime,
                duration: process.burst
            });
            
            // Store process results
            processResults.push({
                id: process.id,
                arrivalTime: process.arrival,
                burstTime: process.burst,
                priority: process.priority,
                startTime: startTime,
                completionTime: completionTime,
                turnaroundTime: turnaroundTime,
                waitingTime: waitingTime,
                responseTime: responseTime
            });
            
            currentTime = completionTime;
        }
        
        // Calculate metrics
        const totalWaitingTime = processResults.reduce((sum, p) => sum + p.waitingTime, 0);
        const totalTurnaroundTime = processResults.reduce((sum, p) => sum + p.turnaroundTime, 0);
        const totalResponseTime = processResults.reduce((sum, p) => sum + p.responseTime, 0);
        const totalBurstTime = processResults.reduce((sum, p) => sum + p.burstTime, 0);
        
        const avgWaitingTime = totalWaitingTime / processResults.length;
        const avgTurnaroundTime = totalTurnaroundTime / processResults.length;
        const avgResponseTime = totalResponseTime / processResults.length;
        const cpuUtilization = totalBurstTime / currentTime;
        
        return {
            algorithm: 'Priority',
            ganttChart: {
                processes: ganttChart,
                totalTime: currentTime
            },
            processResults: processResults,
            metrics: {
                avgWaitingTime: avgWaitingTime,
                avgTurnaroundTime: avgTurnaroundTime,
                avgResponseTime: avgResponseTime,
                cpuUtilization: cpuUtilization
            }
        };
    }
}

File: VizOs2/VizOs/src/algorithms/Priority.py
"""
Priority Scheduling Algorithm
CPU scheduling algorithm based on process priority
"""

def execute(processes, preemptive=False):
    """
    Execute Priority scheduling algorithm
    
    Args:
        processes: List of process dictionaries with priority
        preemptive: Whether to use preemptive scheduling
        
    Returns:
        Dictionary containing gantt_chart, metrics, and process_details
    """
    if preemptive:
        return _execute_preemptive(processes)
    else:
        return _execute_non_preemptive(processes)

def _execute_non_preemptive(processes):
    """Non-preemptive priority scheduling"""
    sorted_processes = sorted(processes, key=lambda x: (x['arrival_time'], x['priority']))
    
    gantt_chart = []
    process_details = []
    current_time = 0
    completed = set()
    
    while len(completed) < len(sorted_processes):
        available = [
            p for p in sorted_processes
            if p['id'] not in completed and p['arrival_time'] <= current_time
        ]
        
        if not available:
            current_time = min(
                p['arrival_time'] for p in sorted_processes
                if p['id'] not in completed
            )
            continue
        
        # Select process with highest priority (lowest number)
        selected = min(available, key=lambda x: x['priority'])
        
        arrival = selected['arrival_time']
        burst = selected['burst_time']
        
        if current_time < arrival:
            current_time = arrival
        
        waiting_time = current_time - arrival
        turnaround_time = waiting_time + burst
        
        gantt_chart.append({
            'process': selected['id'],
            'start': current_time,
            'end': current_time + burst
        })
        
        process_details.append({
            'id': selected['id'],
            'arrival_time': arrival,
            'burst_time': burst,
            'priority': selected['priority'],
            'waiting_time': waiting_time,
            'turnaround_time': turnaround_time,
            'completion_time': current_time + burst
        })
        
        current_time += burst
        completed.add(selected['id'])
    
    # Calculate metrics
    total_waiting_time = sum(p['waiting_time'] for p in process_details)
    total_turnaround_time = sum(p['turnaround_time'] for p in process_details)
    n = len(process_details)
    
    metrics = {
        'average_waiting_time': round(total_waiting_time / n, 2),
        'average_turnaround_time': round(total_turnaround_time / n, 2),
        'cpu_utilization': round((sum(p['burst_time'] for p in process_details) / current_time) * 100, 2) if current_time > 0 else 0,
        'throughput': round(n / current_time, 2) if current_time > 0 else 0
    }
    
    return {
        'gantt_chart': gantt_chart,
        'metrics': metrics,
        'process_details': process_details
    }

def _execute_preemptive(processes):
    """Preemptive priority scheduling (similar to Round Robin but priority-based)"""
    # Simplified implementation - similar to non-preemptive for now
    # Full preemptive implementation would require time-slicing
    return _execute_non_preemptive(processes)

